client.go

```diff
@@ -935,6 +935,46 @@ func (c *Client[TTx]) runProducers(fetchNewWorkCtx, workCtx context.Context) {
 	}
 }
 
+// Cancel cancels the job with the given ID. If possible, the job is cancelled
+// immediately and will not be retried. The provided context is used for the
+// underlying Postgres update and can be used to cancel the operation or apply a
+// timeout.
+//
+// If the job is still in the queue (available, scheduled, or retryable), it is
+// immediately marked as cancelled and will not be retried.
+//
+// If the job is already finalized (cancelled, completed, or discarded), no
+// changes are made.
+//
+// If the job is currently running, it is not immediately cancelled, but is
+// instead marked for cancellation. The client running the job will also be
+// notified (via LISTEN/NOTIFY) to cancel the running job's context. Although
+// the job's context will be cancelled, since Go does not provide a mechanism to
+// interrupt a running goroutine the job will continue running until it returns.
+// As always, it is important for workers to respect context cancellation and
+// return promptly when the job context is done.
+//
+// Once the cancellation signal is received by the client running the job, any
+// error returned by that job will result in it being cancelled permanently and
+// not retried. However if the job returns no error, it will be completed as
+// usual.
+//
+// In the event the running job finishes executing _before_ the cancellation
+// signal is received, the job will be completed or retried as usual without
+// regard to the cancellation. TODO: fix this w/ smarter completion query that
+// uses metadata value?
```

Comment:
See the question in the PR description.

Reply:
IMO: if the job was about to be completed, just let that happen without the cancellation superseding it. However, if a job that was cancelled were to be set as retryable with the possibility of running again, that definitely seems wrong and should be fixed.

Reply:
Yes, I agree. That would require changes only to the queries which would mark the job as errored, because in that case you'd want to realize that the cancellation had been attempted and you'd want that to take precedence.

Reply:
I started working on this, but quickly realized it's a bit more nuanced than I expected. [This query](https://github.com/riverqueue/river/blob/f3fabdff6c2010b93938c0bd1602d8f14ec5fb09/internal/dbsqlc/river_job.sql#L241-L259) is the one making the update from the completer. I wanted to confirm the desired outcome for each of the potential states it could transition into:

* `completed`: this one's easy, just let it complete / no changes
* `cancelled`: also easy, let it cancel / no changes
* `discarded`: since the job is going to finalize and not retry again, I think this one is also fine to not change?
* `retryable`: we don't want additional retries to proceed, so we should (a) set a `finalized_at` and (b) override the `state` to `cancelled`. But we won't have a `@finalized_at` input here because the caller hadn't intended to finalize it, so I guess I'd just use `now()` in the query?
* `scheduled`: I guess similar to above, you would also want to override this one to set a `finalized_at` and a `state` of `cancelled`.

I implemented the above behavior just now, please take a look. I haven't yet added tests because I want to make sure this all sounds reasonable to you first (the tests may be tricky).